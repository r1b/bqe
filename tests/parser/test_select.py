import pytest


@pytest.mark.parametrize(
    "sql",
    (
        pytest.param("SELECT 1", id="select_literal"),
        pytest.param("SELECT column1 FROM table1", id="select_list_single"),
        pytest.param("SELECT column1, column2 FROM table1", id="select_list_multi"),
        pytest.param("SELECT column1 AS alias1 FROM table1", id="select_list_alias_explicit"),
        pytest.param("SELECT column1 alias1 FROM table1", id="select_list_alias_implicit"),
        pytest.param("SELECT column1, FROM table2", id="select_list_trailing_comma"),
        pytest.param("SELECT * FROM table1", id="select_star"),
        pytest.param("SELECT column1.* FROM table1", id="select_star_scoped"),
        pytest.param(
            "SELECT column1.column2.* FROM table1",
            id="select_star_scoped_nested",
        ),
        pytest.param("SELECT column1 FROM table1 AS alias1", id="select_from_alias_explicit"),
        pytest.param("SELECT column1 FROM table1 alias1", id="select_from_alias_implicit"),
        pytest.param(
            "SELECT column1 FROM table1 FOR SYSTEM_TIME AS OF '1970-01-01'",
            id="select_from_time_travel",
        ),
        pytest.param("SELECT AS STRUCT column1 FROM table1", id="select_as_struct"),
        pytest.param("SELECT AS VALUE column1 FROM table1", id="select_as_value"),
        pytest.param(
            """
            SELECT
                *,
                column1,
                (SELECT 1) AS column2
            FROM
                (SELECT column7 FROM table2) AS alias2
            """,
            id="select_subqueries",
        ),
        pytest.param(
            """
            SELECT DISTINCT column1 FROM table1
            """,
            id="select_distinct",
        ),
        pytest.param(
            """
            SELECT * EXCEPT(column1, column2) FROM table1
            """,
            id="select_star_except",
        ),
        pytest.param(
            """
            SELECT struct_column.* EXCEPT(column1, column2) FROM table1
            """,
            id="select_dot_star_except",
        ),
        pytest.param(
            """
            SELECT * REPLACE(1 AS column1, 2 AS column2) FROM table1
            """,
            id="select_star_replace",
        ),
        pytest.param(
            """
            SELECT struct_column.* REPLACE(1 AS column1, 2 AS column2) FROM table1
            """,
            id="select_dot_star_replace",
        ),
    ),
)
def test_select_ok(sql, assert_parse_tree):
    assert_parse_tree(sql)


@pytest.mark.parametrize(
    "sql",
    (
        pytest.param("SELECT ,", id="select_comma_only"),
        pytest.param("SELECT ,1", id="select_comma_leading"),
        # FIXME: BQ says "Syntax error: SELECT list must not be empty at [1:8]"
        pytest.param(
            "SELECT from FROM table1", id="select_column_reserved_word", marks=pytest.mark.xfail
        ),
        pytest.param(
            "SELECT * REPLACE (column1 alias1) FROM table1", id="select_star_replace_alias_implicit"
        ),
    ),
)
def test_select_error(sql, assert_parse_tree_error):
    assert_parse_tree_error(sql)


@pytest.mark.parametrize(
    "sql",
    (
        pytest.param("(SELECT 1)", id="subquery"),
        pytest.param("(SELECT (SELECT (SELECT 1)))", id="subquery_nested"),
        pytest.param("SELECT column1 FROM (SELECT column1 FROM table1)", id="select_from_subquery"),
        pytest.param(
            "SELECT column1 FROM (SELECT column1 FROM table1) AS table2",
            id="select_from_subquery_aliased",
        ),
    ),
)
def test_subquery_ok(sql, assert_parse_tree):
    assert_parse_tree(sql)


@pytest.mark.parametrize(
    "sql",
    (
        pytest.param("SELECT column1 FROM UNNEST(array1)", id="unnest"),
        pytest.param("SELECT column1 FROM UNNEST(array1) AS alias1", id="unnest_alias"),
        pytest.param(
            "SELECT column1 FROM UNNEST(array1) WITH OFFSET",
            id="unnest_with_offset",
        ),
        pytest.param(
            "SELECT column1 FROM UNNEST(array1) WITH OFFSET AS alias1",
            id="unnest_with_offset_alias",
        ),
        pytest.param(
            "SELECT column1 FROM UNNEST(array1) AS alias1 WITH OFFSET AS alias2",
            id="unnest_alias_with_offset_alias",
        ),
    ),
)
def test_unnest_ok(sql, assert_parse_tree):
    assert_parse_tree(sql)
