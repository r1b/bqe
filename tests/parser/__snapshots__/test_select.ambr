# serializer version: 1
# name: test_select_ok[select_as_struct]
  '''
  SELECT AS STRUCT column1 FROM table1
  
  query_expr
    select
      select_metadata
        select_shape	STRUCT
      select_list
        select_expr
          ident	column1
      from_clause
        from_item_table
          path_expression
            ident	table1
          None
          None
  
  Query
    SelectAsStruct(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
      FromClause
        TablePathExpression
          PathExpression
            Ident(table1)
  '''
# ---
# name: test_select_ok[select_as_value]
  '''
  SELECT AS VALUE column1 FROM table1
  
  query_expr
    select
      select_metadata
        select_shape	VALUE
      select_list
        select_expr
          ident	column1
      from_clause
        from_item_table
          path_expression
            ident	table1
          None
          None
  
  Query
    SelectAsValue(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
      FromClause
        TablePathExpression
          PathExpression
            Ident(table1)
  '''
# ---
# name: test_select_ok[select_distinct]
  '''
  SELECT DISTINCT column1 FROM table1
  
  query_expr
    select
      select_metadata
        select_mode	DISTINCT
      select_list
        select_expr
          ident	column1
      from_clause
        from_item_table
          path_expression
            ident	table1
          None
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
      FromClause
        TablePathExpression
          PathExpression
            Ident(table1)
  '''
# ---
# name: test_select_ok[select_dot_star_except]
  '''
  SELECT struct_column.* EXCEPT(column1, column2) FROM table1
  
  query_expr
    select
      select_metadata
      select_list
        select_dot_star
          ident	struct_column
          select_except
            select_except_list
              select_except_item
                ident	column1
              select_except_item
                ident	column2
          None
      from_clause
        from_item_table
          path_expression
            ident	table1
          None
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectDotStar
          Ident(struct_column)
          ExceptModifierList
            ExceptModifierItem
              Ident(column1)
            ExceptModifierItem
              Ident(column2)
      FromClause
        TablePathExpression
          PathExpression
            Ident(table1)
  '''
# ---
# name: test_select_ok[select_dot_star_replace]
  '''
  SELECT struct_column.* REPLACE(1 AS column1, 2 AS column2) FROM table1
  
  query_expr
    select
      select_metadata
      select_list
        select_dot_star
          ident	struct_column
          None
          select_replace
            select_replace_list
              select_replace_item
                integer_literal	1
                ident	column1
              select_replace_item
                integer_literal	2
                ident	column2
      from_clause
        from_item_table
          path_expression
            ident	table1
          None
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectDotStar
          Ident(struct_column)
          ReplaceModifierList
            ReplaceModifierItem
              IntegerLiteral(1)
              Ident(column1)
            ReplaceModifierItem
              IntegerLiteral(2)
              Ident(column2)
      FromClause
        TablePathExpression
          PathExpression
            Ident(table1)
  '''
# ---
# name: test_select_ok[select_from_alias_explicit]
  '''
  SELECT column1 FROM table1 AS alias1
  
  query_expr
    select
      select_metadata
      select_list
        select_expr
          ident	column1
      from_clause
        from_item_table
          path_expression
            ident	table1
          as_alias
            ident	alias1
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
      FromClause
        TablePathExpression
          PathExpression
            Ident(table1)
          Alias
            Ident(alias1)
  '''
# ---
# name: test_select_ok[select_from_alias_implicit]
  '''
  SELECT column1 FROM table1 alias1
  
  query_expr
    select
      select_metadata
      select_list
        select_expr
          ident	column1
      from_clause
        from_item_table
          path_expression
            ident	table1
          as_alias
            ident	alias1
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
      FromClause
        TablePathExpression
          PathExpression
            Ident(table1)
          Alias
            Ident(alias1)
  '''
# ---
# name: test_select_ok[select_from_time_travel]
  '''
  SELECT column1 FROM table1 FOR SYSTEM_TIME AS OF '1970-01-01'
  
  query_expr
    select
      select_metadata
      select_list
        select_expr
          ident	column1
      from_clause
        from_item_table
          path_expression
            ident	table1
          None
          time_travel
            string_literal	'1970-01-01'
  
  Query
    Select(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
      FromClause
        TablePathExpression
          PathExpression
            Ident(table1)
          TimeTravel
            StringLiteral('1970-01-01')
  '''
# ---
# name: test_select_ok[select_list_alias_explicit]
  '''
  SELECT column1 AS alias1 FROM table1
  
  query_expr
    select
      select_metadata
      select_list
        select_expr
          ident	column1
          as_alias
            ident	alias1
      from_clause
        from_item_table
          path_expression
            ident	table1
          None
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
          Alias
            Ident(alias1)
      FromClause
        TablePathExpression
          PathExpression
            Ident(table1)
  '''
# ---
# name: test_select_ok[select_list_alias_implicit]
  '''
  SELECT column1 alias1 FROM table1
  
  query_expr
    select
      select_metadata
      select_list
        select_expr
          ident	column1
          as_alias
            ident	alias1
      from_clause
        from_item_table
          path_expression
            ident	table1
          None
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
          Alias
            Ident(alias1)
      FromClause
        TablePathExpression
          PathExpression
            Ident(table1)
  '''
# ---
# name: test_select_ok[select_list_multi]
  '''
  SELECT column1, column2 FROM table1
  
  query_expr
    select
      select_metadata
      select_list
        select_expr
          ident	column1
        select_expr
          ident	column2
      from_clause
        from_item_table
          path_expression
            ident	table1
          None
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
        SelectColumn
          Ident(column2)
      FromClause
        TablePathExpression
          PathExpression
            Ident(table1)
  '''
# ---
# name: test_select_ok[select_list_single]
  '''
  SELECT column1 FROM table1
  
  query_expr
    select
      select_metadata
      select_list
        select_expr
          ident	column1
      from_clause
        from_item_table
          path_expression
            ident	table1
          None
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
      FromClause
        TablePathExpression
          PathExpression
            Ident(table1)
  '''
# ---
# name: test_select_ok[select_list_trailing_comma]
  '''
  SELECT column1, FROM table2
  
  query_expr
    select
      select_metadata
      select_list
        select_expr
          ident	column1
      from_clause
        from_item_table
          path_expression
            ident	table2
          None
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
      FromClause
        TablePathExpression
          PathExpression
            Ident(table2)
  '''
# ---
# name: test_select_ok[select_literal]
  '''
  SELECT 1
  
  query_expr
    select
      select_metadata
      select_list
        select_expr
          integer_literal	1
  
  Query
    Select(distinct=False)
      SelectList
        SelectColumn
          IntegerLiteral(1)
  '''
# ---
# name: test_select_ok[select_star]
  '''
  SELECT * FROM table1
  
  query_expr
    select
      select_metadata
      select_list
        select_star
          None
          None
      from_clause
        from_item_table
          path_expression
            ident	table1
          None
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectStar
      FromClause
        TablePathExpression
          PathExpression
            Ident(table1)
  '''
# ---
# name: test_select_ok[select_star_except]
  '''
  SELECT * EXCEPT(column1, column2) FROM table1
  
  query_expr
    select
      select_metadata
      select_list
        select_star
          select_except
            select_except_list
              select_except_item
                ident	column1
              select_except_item
                ident	column2
          None
      from_clause
        from_item_table
          path_expression
            ident	table1
          None
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectStar
          ExceptModifierList
            ExceptModifierItem
              Ident(column1)
            ExceptModifierItem
              Ident(column2)
      FromClause
        TablePathExpression
          PathExpression
            Ident(table1)
  '''
# ---
# name: test_select_ok[select_star_replace]
  '''
  SELECT * REPLACE(1 AS column1, 2 AS column2) FROM table1
  
  query_expr
    select
      select_metadata
      select_list
        select_star
          None
          select_replace
            select_replace_list
              select_replace_item
                integer_literal	1
                ident	column1
              select_replace_item
                integer_literal	2
                ident	column2
      from_clause
        from_item_table
          path_expression
            ident	table1
          None
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectStar
          ReplaceModifierList
            ReplaceModifierItem
              IntegerLiteral(1)
              Ident(column1)
            ReplaceModifierItem
              IntegerLiteral(2)
              Ident(column2)
      FromClause
        TablePathExpression
          PathExpression
            Ident(table1)
  '''
# ---
# name: test_select_ok[select_star_scoped]
  '''
  SELECT column1.* FROM table1
  
  query_expr
    select
      select_metadata
      select_list
        select_dot_star
          ident	column1
          None
          None
      from_clause
        from_item_table
          path_expression
            ident	table1
          None
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectDotStar
          Ident(column1)
      FromClause
        TablePathExpression
          PathExpression
            Ident(table1)
  '''
# ---
# name: test_select_ok[select_star_scoped_nested]
  '''
  SELECT column1.column2.* FROM table1
  
  query_expr
    select
      select_metadata
      select_list
        select_dot_star
          path_expression
            ident	column1
            ident	column2
          None
          None
      from_clause
        from_item_table
          path_expression
            ident	table1
          None
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectDotStar
          PathExpression
            Ident(column1)
            Ident(column2)
      FromClause
        TablePathExpression
          PathExpression
            Ident(table1)
  '''
# ---
# name: test_select_ok[select_subqueries]
  '''
  SELECT
      *,
      column1,
      (SELECT 1) AS column2
  FROM
      (SELECT column7 FROM table2) AS alias2
  
  query_expr
    select
      select_metadata
      select_list
        select_star
          None
          None
        select_expr
          ident	column1
        select_expr
          subquery
            query_expr
              select
                select_metadata
                select_list
                  select_expr
                    integer_literal	1
          as_alias
            ident	column2
      from_clause
        from_item_subquery
          subquery
            query_expr
              select
                select_metadata
                select_list
                  select_expr
                    ident	column7
                from_clause
                  from_item_table
                    path_expression
                      ident	table2
                    None
                    None
          as_alias
            ident	alias2
  
  Query
    Select(distinct=False)
      SelectList
        SelectStar
        SelectColumn
          Ident(column1)
        SelectColumn
          Subquery
            Query
              Select(distinct=False)
                SelectList
                  SelectColumn
                    IntegerLiteral(1)
          Alias
            Ident(column2)
      FromClause
        TableSubquery
          Subquery
            Query
              Select(distinct=False)
                SelectList
                  SelectColumn
                    Ident(column7)
                FromClause
                  TablePathExpression
                    PathExpression
                      Ident(table2)
          Alias
            Ident(alias2)
  '''
# ---
# name: test_subquery_ok[select_from_subquery]
  '''
  SELECT column1 FROM (SELECT column1 FROM table1)
  
  query_expr
    select
      select_metadata
      select_list
        select_expr
          ident	column1
      from_clause
        from_item_subquery
          subquery
            query_expr
              select
                select_metadata
                select_list
                  select_expr
                    ident	column1
                from_clause
                  from_item_table
                    path_expression
                      ident	table1
                    None
                    None
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
      FromClause
        TableSubquery
          Subquery
            Query
              Select(distinct=False)
                SelectList
                  SelectColumn
                    Ident(column1)
                FromClause
                  TablePathExpression
                    PathExpression
                      Ident(table1)
  '''
# ---
# name: test_subquery_ok[select_from_subquery_aliased]
  '''
  SELECT column1 FROM (SELECT column1 FROM table1) AS table2
  
  query_expr
    select
      select_metadata
      select_list
        select_expr
          ident	column1
      from_clause
        from_item_subquery
          subquery
            query_expr
              select
                select_metadata
                select_list
                  select_expr
                    ident	column1
                from_clause
                  from_item_table
                    path_expression
                      ident	table1
                    None
                    None
          as_alias
            ident	table2
  
  Query
    Select(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
      FromClause
        TableSubquery
          Subquery
            Query
              Select(distinct=False)
                SelectList
                  SelectColumn
                    Ident(column1)
                FromClause
                  TablePathExpression
                    PathExpression
                      Ident(table1)
          Alias
            Ident(table2)
  '''
# ---
# name: test_subquery_ok[subquery]
  '''
  (SELECT 1)
  
  query_expr
    subquery
      query_expr
        select
          select_metadata
          select_list
            select_expr
              integer_literal	1
  
  Query
    Subquery
      Query
        Select(distinct=False)
          SelectList
            SelectColumn
              IntegerLiteral(1)
  '''
# ---
# name: test_subquery_ok[subquery_nested]
  '''
  (SELECT (SELECT (SELECT 1)))
  
  query_expr
    subquery
      query_expr
        select
          select_metadata
          select_list
            select_expr
              subquery
                query_expr
                  select
                    select_metadata
                    select_list
                      select_expr
                        subquery
                          query_expr
                            select
                              select_metadata
                              select_list
                                select_expr
                                  integer_literal	1
  
  Query
    Subquery
      Query
        Select(distinct=False)
          SelectList
            SelectColumn
              Subquery
                Query
                  Select(distinct=False)
                    SelectList
                      SelectColumn
                        Subquery
                          Query
                            Select(distinct=False)
                              SelectList
                                SelectColumn
                                  IntegerLiteral(1)
  '''
# ---
# name: test_unnest_ok[unnest]
  '''
  SELECT column1 FROM UNNEST(array1)
  
  query_expr
    select
      select_metadata
      select_list
        select_expr
          ident	column1
      from_clause
        from_item_unnest
          unnest_expr
            ident	array1
          None
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
      FromClause
        TableUnnestExpression
          UnnestExpression
            Ident(array1)
  '''
# ---
# name: test_unnest_ok[unnest_alias]
  '''
  SELECT column1 FROM UNNEST(array1) AS alias1
  
  query_expr
    select
      select_metadata
      select_list
        select_expr
          ident	column1
      from_clause
        from_item_unnest
          unnest_expr
            ident	array1
          as_alias
            ident	alias1
          None
  
  Query
    Select(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
      FromClause
        TableUnnestExpression
          UnnestExpression
            Ident(array1)
          Alias
            Ident(alias1)
  '''
# ---
# name: test_unnest_ok[unnest_alias_with_offset_alias]
  '''
  SELECT column1 FROM UNNEST(array1) AS alias1 WITH OFFSET AS alias2
  
  query_expr
    select
      select_metadata
      select_list
        select_expr
          ident	column1
      from_clause
        from_item_unnest
          unnest_expr
            ident	array1
          as_alias
            ident	alias1
          from_unnest_offset_expr
            as_alias
              ident	alias2
  
  Query
    Select(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
      FromClause
        TableUnnestExpression
          UnnestExpression
            Ident(array1)
          Alias
            Ident(alias1)
          WithOffset
            Alias
              Ident(alias2)
  '''
# ---
# name: test_unnest_ok[unnest_with_offset]
  '''
  SELECT column1 FROM UNNEST(array1) WITH OFFSET
  
  query_expr
    select
      select_metadata
      select_list
        select_expr
          ident	column1
      from_clause
        from_item_unnest
          unnest_expr
            ident	array1
          None
          from_unnest_offset_expr	None
  
  Query
    Select(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
      FromClause
        TableUnnestExpression
          UnnestExpression
            Ident(array1)
          WithOffset
  '''
# ---
# name: test_unnest_ok[unnest_with_offset_alias]
  '''
  SELECT column1 FROM UNNEST(array1) WITH OFFSET AS alias1
  
  query_expr
    select
      select_metadata
      select_list
        select_expr
          ident	column1
      from_clause
        from_item_unnest
          unnest_expr
            ident	array1
          None
          from_unnest_offset_expr
            as_alias
              ident	alias1
  
  Query
    Select(distinct=False)
      SelectList
        SelectColumn
          Ident(column1)
      FromClause
        TableUnnestExpression
          UnnestExpression
            Ident(array1)
          WithOffset
            Alias
              Ident(alias1)
  '''
# ---
